src/WCSimConstructCylinder.cc

```diff
@@ -1307,8 +1666,159 @@ G4LogicalVolume* WCSimDetectorConstruction::ConstructCaps(G4int zflip)
+                              PMTPosition,
+                              logicWCPMTOD,                // its logical volume
+                              "WCPMTOD",             // its name
+                              logicWCBarrelBorderCell,         // its mother volume
+                              false,                     // no boolean operations
+                              (int)(i*WCPMTODperCellVertical+j),
+                              true);
+
+      }
+    }
+
+    if(!(WCBarrelRingNPhi*WCPMTperCellHorizontal == WCBarrelNumPMTHorizontal)){
+
+      G4VPhysicalVolume* physiWCExtraBorderODTyvek =
+        new G4PVPlacement(0,
+                              G4ThreeVector(0.,0.,0.),
+                              logicWCTowerODTyvek,
+                              "WCExtraTowerODTyvek",
+                              logicWCExtraBorderCell,
+                              false,
+                              0,true);
+
+			G4LogicalBorderSurface * WaterTyBarrelBorderCellSurfaceBot =
+					new G4LogicalBorderSurface(	"WaterTyBarrelBorderCellSurface",
+																			 physiWCExtraBorderCell,
+																			 physiWCExtraBorderODTyvek,
+																			 OpWaterTySurface);
+
+			G4RotationMatrix* WCPMTRotation = new G4RotationMatrix;
+      WCPMTRotation->rotateY(270.*deg);
+      WCPMTRotation->rotateX((2*pi-totalAngle)/2.);//align the PMT with the Cell
+
+      G4double towerWidthOD = WCODRadius*tan(2*pi-totalAngle);
+
+      G4double horizontalODSpacing = towerWidthOD/WCPMTODperCellHorizontal;
+      G4double verticalODSpacing   = barrelCellHeight*(WCODRadius/WCIDRadius)/WCPMTODperCellVertical;
+
+      for(G4double i = 0; i < (WCPMTODperCellHorizontal); i++){
+        for(G4double j = 0; j < WCPMTODperCellVertical; j++){
+          G4ThreeVector PMTPosition =  G4ThreeVector(WCODRadius,
+                                                     -towerWidthOD/2.+(i+0.5)*horizontalODSpacing,
+                                                     -(barrelCellHeight * (WCODRadius/WCIDRadius))/2.+(j+0.5)*verticalODSpacing);
+          PMTPosition.rotateZ(-(2*pi-totalAngle)/2.); // align with the symmetry
+          //axes of the cell
+
+
+          G4VPhysicalVolume* physiWCBarrelPMT =
+              new G4PVPlacement(WCPMTRotation,             // its rotation
+                                PMTPosition,
+                                logicWCPMTOD,                // its logical volume
+                                "WCPMTOD",             // its name
+                                logicWCExtraBorderCell,         // its mother volume
+                                false,                     // no boolean operations
+                                (int)(i*WCPMTODperCellVertical+j),
+                                true);
+
+        }
+      }
+
+    }
+
+  } // END isODConstructed
+
+  return logicCapAssembly;
 
 }
 
 
+void ComputeWCODPMT(G4int NPMT, G4double *NPMTHorizontal, G4double *NPMTVertical){
+  switch (NPMT) {
```

Comment:
Why is this function needed?

Reply:
This function computes the nb of PMTs on the barrel side if the user decides to not define WCPMTODperCellHorizontal and WCPMTODperCellVertical by setting them to 0. Therefore one has the possibility to only define a coverage and WCSim will compute the nb of PMTs to create itself. I added a description inside WCSimDetectorConfigs, SetHyperKWithODGeometry() method.